### ISSUE TITLE
"az storage message get" crashes (Ubuntu 20.04 LTS, az-cli 2.36.0-1~focal)

### ISSUE THREAD

**Issue by hholst80** [Azure/azure-cli#22268]
**Describe the bug**

az storage queue get fails

**To Reproduce**

1. start azurite (or connect to a azure hosted storage account)
2. set AZURE_STORAGE_CONNECTION_STRING
3. az storage queue create --name egress-test
4. az storage message get -q "egress-test" 

**Expected behavior**

This should not crash.

```
❯   az storage message get -q "egress-test"   
Command group 'storage message' is in preview and under development. Reference and support levels: https://aka.ms/CLI_refstatus
The command failed with an unexpected error. Here is the traceback:
End of paging
Traceback (most recent call last):                                                                                                                            
  File "/opt/az/lib/python3.8/site-packages/knack/cli.py", line 231, in invoke                                                                                
    cmd_result = self.invocation.execute(args)                                                                                                                
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/__init__.py", line 658, in execute                                                        
    raise ex                                                                                                                                                  
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/__init__.py", line 721, in _run_jobs_serially                                             
    results.append(self._run_job(expanded_arg, cmd_copy))                                                                                                     
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/__init__.py", line 713, in _run_job                                                       
    return cmd_copy.exception_handler(ex)                                                                                                                     
  File "/opt/az/lib/python3.8/site-packages/azure/cli/command_modules/storage/__init__.py", line 409, in new_handler                                          
    first(ex)                                                                                                                                                 
  File "/opt/az/lib/python3.8/site-packages/azure/cli/command_modules/storage/__init__.py", line 408, in new_handler                                          
    raise ex                                                                                                                                                  
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/__init__.py", line 692, in _run_job                                                       
    result = cmd_copy(params)                                                                                                                                 
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/__init__.py", line 328, in __call__                                                       
    return self.handler(*args, **kwargs)                                                                                                                      
  File "/opt/az/lib/python3.8/site-packages/azure/cli/core/commands/command_operation.py", line 121, in handler                                               
    return op(**command_args)                                                                                                                                 
  File "/opt/az/lib/python3.8/site-packages/azure/cli/command_modules/storage/operations/queue.py", line 86, in receive_messages                              
    return list(next(page_iter))                                                                                                                              
  File "/opt/az/lib/python3.8/site-packages/azure/core/paging.py", line 84, in __next__                                                                       
    self.continuation_token, self._current_page = self._extract_data(self._response)                                                                          
  File "/opt/az/lib/python3.8/site-packages/azure/multiapi/storagev2/queue/v2018_03_28/_models.py", line 288, in _extract_data_cb                             
    raise StopIteration("End of paging")                                                                                                                      
StopIteration: End of paging                                                                                                                                  
To open an issue, please run: 'az feedback'
```

**Environment summary**

```
ubuntu 20.04 container with azure-cli installed from ms repo.

ii  azure-cli                  2.36.0-1~focal                    all          Azure CLI
```

**Comment by hholst80** [comment]
azure-cli==2.34.1 worked, seems to have been broken since.

**Comment by yonzhan** [comment]
storage

**Comment by evelyn-ys** [comment]
@jonie001 Can you check if it's caused by track2 migration?

**Comment by jonie001** [comment]
@tasherif-msft @evelyn-ys @yonzhan  This error occurs when message is empty, call receive_messages module in the azure-storage-queue.

**Comment by evelyn-ys** [comment]
Route to Storage Python SDK team

**Comment by hholst80** [comment]
> @tasherif-msft This error occurs when message is empty, call receive_messages module in the azure-storage-queue.

I saw it for non-empty messages too in our CI/CD environment. The repro is for an empty queue, but that is just one way to trigger the error condition.

**Comment by navba-MSFT** [comment]
@hholst80 Thanks for reaching out to us and sharing this feedback. I am able to reproduce this issue at my end with 2.36.0 version of Azure CLI. However, when I tried with 2.35.0 version it worked just fine.  

The failure exception is raised here:

Source code: https://github.com/Azure/azure-sdk-for-python/blob/79c58f79c1168b35a804c2d6ca4977db50968819/sdk/storage/azure-storage-queue/azure/storage/queue/_models.py#L366

```
def _extract_data_cb(self, messages): # pylint: disable=no-self-use
        # There is no concept of continuation token, so raising on my own condition
        if not messages:
            raise StopIteration("End of paging")    #----- HERE ---- 
        if self._max_messages is not None:
            self._max_messages = self._max_messages - len(messages)
        return "TOKEN_IGNORED", [QueueMessage._from_generated(q) for q in messages]  # pylint: disable=protected-access
```

I don’t see any changes being done to the queue cli command from 2.35 to 2.36. See [release notes](https://docs.microsoft.com/en-us/cli/azure/release-notes-azure-cli#storage). 

I am looking into this further and I will provide an update once I have more details.

**Comment by jalauzon-msft** [comment]
Hi @evelyn-ys Yishi / @navba-MSFT, I don't believe this to be an SDK issue but rather an issue with how the CLI is calling the Track2 `receive_messages` API. Here is the code snippet I think CLI is using:

https://github.com/Azure/azure-cli/blob/cc0d149c1d3bea7ffbac12f8e97e0286cddfaa23/src/azure-cli/azure/cli/command_modules/storage/operations/queue.py#L84-L86

When using `by_page()` and then calling `next` on the resulting Iterator, it is expected that the Iterator will throw a `StopIteration` exception if there is not a next item, like is the case when the Queue is empty. There are a couple of immediate ways around this that I can think of: Catching the `StopIteration` exception or using the default value param of `next()`.

```python
# Option 1: Default value for next() to empty list
page_iter = client.receive_messages(**kwargs).by_page()
return list(next(page_iter, []))  # Maybe split lines out to avoid extra list construction

# Option 2: Catch and handle StopIteration exception
page_iter = client.receive_messages(**kwargs).by_page()
try:
    page = next(page_iter)
except StopIteration:
    return []
return list(page)
```

Either of those should solve the Issue at hand but I'm also curious what the CLI actually is trying to do? Is this intention to receive a certain number of messages? Is that user settable?
- If you just need a single message, I would use `receive_message` singular from the SDK
- If you need a user-specified amount, there are two approaches:
  - The approach that is done today should work. I assume the user can set `messages_per_page` or something alike? There is some odd behavior with `receive_messages` in this case where it will continually get messages unless you ask for each page via `by_page()` so the approach is correct here.
  - We just introduced a `max_messages` parameter to this function (in our next release) where you can specify a maximum number of messages so it does not continually grab messages from the Queue. In that case you would not need to use `by_page()` at all and could just do something like below, which would also eliminate the `StopIteration` exception on an empty Queue. (Also see the [SDK sample](https://github.com/Azure/azure-sdk-for-python/blob/7ff526a2f5553079a4cbdf8a76e03f8338fbef56/sdk/storage/azure-storage-queue/samples/queue_samples_message.py#L303))
 
```python
pages = client.receive_messages(max_messages=5, **kwargs)
return list(pages)
```
Note: An optimization to the above would be to also set `messages_per_page` since that controls how many messages the server returns in a single request. If not set, the default is 1.

Sorry for the long message but hopefully that clears up the Issue and provides some potential fixes. Please ping me offline if you need more help. Thanks!

**Comment by evelyn-ys** [comment]
Thanks for the detailed explanation @jalauzon-msft.
I'm curious why not `return None, []` but raise exception when messages is `[]`?
```
    def _extract_data_cb(self, messages): # pylint: disable=no-self-use
        # There is no concept of continuation token, so raising on my own condition
        if not messages:
            raise StopIteration("End of paging")
        return "TOKEN_IGNORED", [QueueMessage._from_generated(q) for q in messages]  # pylint: disable=protected-access
```

**Comment by navba-MSFT** [comment]
@evelyn-ys Thanks for your reply.

@jalauzon-msft Could you please share your thoughts on the above comment ?

**Comment by jalauzon-msft** [comment]
@evelyn-ys Yishi, a `StopIteration` exception is the standard way for an iterator to signal the end of the iteration. See [Python docs](https://docs.python.org/3/glossary.html#term-iterator) on iterator. Built-in elements like the `for` loop internally handle the `StopIteration` exception to know when to break the loop. If we were to return an empty list `[]` the iterator would indefinitely return empty lists (when used in a for-loop or similar).

So, when using the iterator directly, via `next()`, a caller must handle the `StopIteration` exception in some form (like the examples I gave above). The general recommendation is to use other, built-in iteration techniques like loops to avoid this but in the case of Queue, accessing the iterator directly is needed if you want to only grab a single page of messages. This is because of the behavior of `recevie_messages` where it will continually grab messages until the Queue is empty if you use normal iteration. Unless you utilize the new `max_messsages` parameter, then looping or `list` constructor could be used as I shared in my example.

As for why this doesn't happen in Blob/file-share, I'd have to see our code and the CLI code to know exactly what's happening but there are cases where those iterators can throw `StopIteration` as well. Though, since they don't face the same issue as Queue, normal iteration, like a `list` constructor can be used to get all the elements without worrying about the exception.
